src/cljx/schema/core.cljx

```diff
@@ -473,47 +422,71 @@
 
 (clojure.core/defrecord Either [schemas]
   Schema
-  (walker [this]
-          (let [sub-walkers (mapv subschema-walker schemas)]
-            (clojure.core/fn [x]
-              (loop [sub-walkers (seq sub-walkers)]
-                (if-not sub-walkers
-                  (macros/validation-error
-                   this x
-                   (list 'some (list 'check '% (utils/value-name x)) 'schemas))
-                  (let [res ((first sub-walkers) x)]
-                    (if-not (utils/error? res)
-                      res
-                      (recur (next sub-walkers)))))))))
+  (spec [this]
+    (variant/variant-spec
+     spec/+no-precondition+
+     (for [s schemas]
+       {:guard (complement (checker s))
```

Comment:
Because the semantics of `variant` when checking is that you use the guard to select the schema, then the value **must** match that schema -- no backtracking.  A 'better' alternative would be to create another custom spec type for `either`, but i wasn't really inclined to make `either` more attractive :)